File: course1/week2/3_greatest_common_divisor/java/src/Main.java
/**
 *
 * Java implementation to find the greatest common divisor between two numbers
 *
 * (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
 *
 **/

import java.util.Scanner;

public class Main {

    private static int gcd( int a, int b ){
        for( int t = 0; b != 0; ){
            t = b;
            b = a % b;
            a = t;
        }
        return a;
    }

    public static void main(String[] args){
        Scanner input = new Scanner( System.in );
        int a = input.nextInt(),
            b = input.nextInt(),
            ans = gcd( a, b );
        System.out.println( ans );
    }
}

File: course1/week2/3_greatest_common_divisor/cpp/main.cpp
/**
 *
 * C implementation to calculate the greatest common divisor between two numbers
 *
 * (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
 *
 **/

#include <iostream>

using namespace std;

template< typename Type >
class Solution {
public:
    Type gcd( Type a, Type b ){
        for( auto t{ b }; b != 0; ){
            t = b;
            b = a % b;
            a = t;
        }
        return a;
    }
};

int main(){
    using Type = size_t;
    Solution< Type > solution;
    Type a = 0,
         b = 0;
    cin >> a >> b;
    Type ans = solution.gcd( a, b );
    cout << ans << endl;
    return 0;
}
